experimental/Schemes/src/SpectralSequences.jl

```diff
@@ -0,0 +1,793 @@
+    produce_lifted_kernel_generators(cssp, i, j)
+  end
+end
+
+function produce_entry(cssp::CSSPage, i::Int, j::Int)
+  is_one(page_number(cssp)) && return produce_entry_on_initial_page(cssp, i, j)
+  css = spectral_sequence(cssp)
+  p = page_number(cssp)
+  previous_page = css[p-1]
+  H = previous_page[i, j]
+  #is_zero(H) && return H
+  B, inc_B = can_compute_map(previous_page, i+p-1, j-p+2) ? image(map(previous_page, i+p-1, j-p+2)) : sub(H, elem_type(H)[])
+  Z, inc_Z = can_compute_map(previous_page, i, j) ? kernel(map(previous_page, i, j)) : sub(H, gens(H))
+  # The code below relies on a particular generating set for the modulus
+  # so we need to carefully build that here.
+  F = ambient_free_module(H)
+  new_B, _ = sub(F, vcat(relations(H), repres.(gens(B))))
+  g = ambient_representatives_generators(Z)
+  return SubquoModule(ambient_free_module(H), g, ambient_representatives_generators(new_B))
+end
+
+function produce_entry_on_initial_page(cssp::CSSPage, i::Int, j::Int)
+  c = graded_complex(cssp)
+  F = c[i]
+  degs = degrees_of_generators(F)
+  ctx = pushforward_ctx(cssp)
+  kk = coefficient_ring(graded_ring(cssp))
+  summands = FreeMod{elem_type(kk)}[cohomology_model(ctx, -d)[j] for d in degs]
+  is_empty(summands) && return FreeMod(kk, 0)
+  return direct_sum(summands)[1]
+end
+
+function can_compute_index(cssp::CSSPage, i::Int, j::Int)
+  can_compute_index(graded_complex(cssp), i) || return false
+  j <= 0 || return false
+  S = graded_ring(cssp)
+  -j <= ngens(S) - rank(grading_group(S)) || return false
+  return true
+end
+
+function can_compute_map(cssp::CSSPage, i::Int, j::Int)
+  p = page_number(cssp)
+  return can_compute_index(cssp, i, j) && can_compute_index(cssp, i-p, j+p-1)
+end
+
+function maps(cssp::CSSPage)
+  if !isdefined(cssp, :maps)
+    cssp.maps = Dict{Tuple{Int, Int}, ModuleFPHom}()
+  end
+  return cssp.maps
+end
+
+function map(cssp::CSSPage, i::Int, j::Int)
+  return get!(maps(cssp), (i, j)) do
+    produce_map(cssp, i, j)
+  end
+end
+
+function produce_lifted_kernel_generators_on_initial_page(cssp::CSSPage, i::Int, j::Int)
+  S = graded_ring(cssp)
+  ctx = pushforward_ctx(cssp)
+  cplx = graded_complex(cssp)
+  orig_map = map(cplx, i+1)
+  dom_degs = degrees_of_generators(cplx[i+1])
+  dom = cssp[i+1, j]
+  result = Tuple{Vector{Int}, Vector{Tuple{Int, FreeModElem}}}[]
+  is_zero(dom) && return result
+  rngs = get_attribute(dom, :ranges)::Vector{UnitRange{Int}}
+  cur_rng_ind = 1
+  summands = get_attribute(dom, :direct_product)::Vector{typeof(dom)}
+  for (v_num, v) in enumerate(gens(dom))
+    buckets = Tuple{Int, FreeModElem}[]
+    e = Int[]
+    ind, _ = only(coordinates(v))
+    while !(ind in rngs[cur_rng_ind])
+      cur_rng_ind+=1
+    end
+    k = cur_rng_ind # findfirst(ind in r for r in rngs)
+    # manual way to project (saves allocations)
+    vv = FreeModElem(coordinates(v)[rngs[cur_rng_ind]], summands[cur_rng_ind]) #canonical_projection(dom, k)(v)
```

Comment:
This should work also in 1.6? (I didn't test the code)
```suggestion
    k = findfirst(r -> ind in r, rngs)
    # manual way to project (saves allocations)
    vv = FreeModElem(coordinates(v)[rngs[k]], summands[k]) #canonical_projection(dom, k)(v)
```

Reply:
Thanks for having a look! However, the point here was not to start looking for `k` from scratch every time, since the `ind` come in in an ordered way. Note that `cur_rng_ind` is another running variable in this loop. This was actually crucial for speed at some point (O(n) vs. O(n^2)).